src/ocean/application/components/GCStats.d

```diff
@@ -0,0 +1,216 @@
+
+        **********************************************************************/
+
+        private alias long MonitorInt;
+    }
+
+    /**********************************************************************
+
+        Structure representing recorded GC stats.
+
+    **********************************************************************/
+
+    public struct Stats
+    {
+        /**********************************************************************
+
+            The number of microseconds for how much the garbage collector was
+            run in total.
+
+        **********************************************************************/
+
+        public size_t gc_run_duration;
+
+        /**********************************************************************
+
+            The percentage of time that was spent by the garbage collector.
+
+        **********************************************************************/
+
+        public float gc_run_percentage = 0;
+
+        /**********************************************************************
+
+            The number of bytes freed overall
+
+        **********************************************************************/
+
+        int freed;
+
+        /**********************************************************************
+
+            The number of bytes freed within full pages
+
+        **********************************************************************/
+
+        int pagebytes;
+    }
+
+    /**********************************************************************
+
+        ditto
+
+    **********************************************************************/
+
+    private Stats stats;
+
+    /**********************************************************************
+
+        Timestamp for when the last garbage collection started
+
+    **********************************************************************/
+
+    private ulong gc_start_us;
+
+    /**********************************************************************
+
+        Timestamp for when the last stats were collected
+
+    **********************************************************************/
+
+    private ulong last_collected_timestamp;
+
+    /**********************************************************************
+
+        Reset statistics.
+
+    **********************************************************************/
+
+    private void reset ( )
+    {
+        this.last_collected_timestamp = MicrosecondsClock.now_us();
+        this.stats.gc_run_duration = 0;
+        this.stats.freed = 0;
+        this.stats.pagebytes = 0;
+    }
+
+    /**********************************************************************
+
+        Start stats reporting
+
+    **********************************************************************/
+
+    public void start ( )
+    {
+        static if ( hasMember!(GC, "monitor") )
+        {
+            GC.monitor(&this.gcBegin, &this.gcEnd);
+        }
```

Comment:
Should we output a warning or something if `hasMember!(GC, "monitor")` is false?

Reply:
Hmm, it's an interesting scenario. The small danger is that an app is happily calling `collect` and logging the GC stats, but in reality no stats are actually being collected, due to the `static if` here not being compiled in. That's only possible in the case of someone using the wrong DMD version, though, so is not a major risk. I'd say your approach of just ignoring this case seems fine.